camera_models.py

```diff
@@ -1129,4 +1130,171 @@ def get_rays_wrt_lidar_frame(self):
+        '''
+        Arguments:
+        pixel_coor (Tensor): A 2xN Tensor contains the pixel coordinates.
+        
+        NOTE: pixel_coor can also have a dimension of Bx2xN, where B is the 
+        batch number.
+        
+        Returns:
+        ray: A 3xN Tensor representing the 3D rays. Bx3xN if batched.
+        valid_mask: A (N,) Tensor representing the valid mask. BxN if batched.
+        '''
+        
+        pixel_coor = self.in_wrap(pixel_coor) # Bx2xN
+        
+        # mx and my becomes float64 if pixel_coor.dtype is integer type.
+        # Convert pixel to angle on the unit sphere.
+
+        mx = ( pixel_coor[..., 0, :] - self.ss.W / 2)
+        my = ( pixel_coor[..., 1, :] - self.ss.H / 2)
+        r = (mx**2.0 + my**2.0)**0.5
+
+        # Compute theta and phi. Theta is the angle of 3D points from the z-axis. Phi is the angle of 3D points from the x-axis on the x-y plane.
+        th = torch.atan2( my, mx )
+        ph = r * self.fov_rad / self.ss.W
+
+        # Valid rays are only those within the field of view.
+        valid_mask = torch.abs(ph) <= self.fov_rad / 2.0
+
+        # Compute the 3D points. Those are of unit length.
+        x = torch.sin(ph) * torch.cos(th)
+        y = torch.sin(ph) * torch.sin(th)
+        z = torch.cos(ph)
+
+        # Need to deal with batch dim
+        ray = torch.stack( (x, y, z), dim=-2 )
+
+        return self.out_wrap(ray), self.out_wrap(valid_mask)
+
+    def point_3d_2_pixel(self, point_3d, normalized=False):
+        '''
+        Arguments:
+        point_3d (Tensor): A 3xN Tensor contains 3D point coordinates. 
+        normalized (bool): If True, then the returned coordinates are normalized to [-1, 1]
+        
+        NOTE: point_3d can also have a dimension of Bx3xN, where B is the 
+        batch number. 
+        
+        Returns: 
+        pixel_coor: A 2xN Tensor representing the 2D pixels. Bx2xN if batched.
+        valid_mask: A (N,) Tensor representing the valid mask. BXN if batched.
+        '''
+
+        point_3d = self.in_wrap(point_3d)
+
+        # torch.split results in Bx1XN.
+        x, y, z = torch.split( point_3d, 1, dim=-2 )
+
+        x2 = x**2.0 # Note: this may promote x2 to torch.float64 if point_3d.dtype=torch.int. 
+        y2 = y**2.0
+        z2 = z**2.0
+
+        # Magnitude of the 3D point.
+        R = (x2 + y2 + z2)**0.5
+
+        # Theta (th) is the angle between the ray and the z-axis.
+        # Phi (ph) is the angle between the ray and the x-axis, in the x-y plane.
+        ph = torch.acos( z / R )
+        th = torch.atan2( y, x )
+
+
+        # The radius in the image plane.
+        r = ph * self.ss.W / self.fov_rad
+
+        # The pixel coordinates.
+        # If any are FTensors, then convert to Tensor.
+        if isinstance(th, FTensor):
+            th = th.tensor()
+            ph = ph.tensor()
+
+
+        u = r * torch.cos(th) + self.ss.W / 2
+        v = r * torch.sin(th) + self.ss.H / 2
+

```

Comment:
@huyaoyu What do you think about this?
I added this piece, where FTensors are converted to torch Tensors, since FTensors do not allow element-wise multiplications.
Does it make sense to do it like this? It feels a bit hacky to me.

Reply:
This is in fact the correct way!

I force the user of ftensor to be absolutely explicit about what they want to do. That is, if two ftensors get multiplied, then it must mean a transform. If it is not a transform, `.tensor()` should not be used to clearly signal the developer her/himself and future developers that the multiplication breaks the state of being an ftensor. They need to handle the frame definitions by themselves then.

Reply:
Awesome!